doc/transfer_secuwall.1
.\"     Title: transfer_secuwall
.\"    Author: 
.\" Generator: DocBook XSL Stylesheets v1.73.2 <http://docbook.sf.net/>
.\"      Date: 02/12/2010
.\"    Manual: 
.\"    Source: 
.\"
.TH "TRANSFER_SECUWALL" "1" "02/12/2010" "" ""
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.SH "NAME"
transfer_secuwall \- secunet wall configuration export utility
.SH "SYNOPSIS"
\fBtransfer_secuwall\fR [\-l] [\-h] [\-n] \-v \fIvolumeid\fR [\-f \fIfilename\&.xml\fR] [\-d \fIworkdir\fR] [\-a \fItemplatedir\fR] \fIfirewall_object_name\fR
.sp
The switches \-a, \-f and \-d are optional\&. If they are not specified, the appropriate defaults are used\&.
.sp
.SH "DESCRIPTION"
transfer_secuwall(1) is a helper utility to compress and export secunet wall host OS specific firewall configuration to a portable device\&.
.sp
It is also capable of searching and displaying all transfer devices of a system which are suitable for config transfer\&.
.sp
.SH "OPTIONS"
.PP
\fB\-l\fR
.RS 4
List all portable devices of the system\&.
.RE
.PP
\fB\-h\fR
.RS 4
Display help text\&.
.RE
.PP
\fB\-n\fR
.RS 4
Append firewall object name to transfer tarball\&. The default is
\fBfalse\fR\&.
.RE
.PP
\fB\-v\fR
.RS 4
Transfer partition\&. Specifies the destination partition for firewall configuration export (e\&.g\&. /dev/sdc1)\&.
.RE
.PP
\fB\-f\fR
.RS 4
Firewall Builder XML file with object definition of firewall to export config\&. If not specified, the filename will be constructed from the
\fIworkdir\fR
and
\fIfirewall_object_name\fR
values:
\fIworkdir\fR
+
\fIfwobjectname\fR
+ \&.fwb
.RE
.PP
\fB\-d\fR
.RS 4
Defines the working directory\&. If not specified, the current directory will be used\&.
.RE
.PP
\fB\-a\fR
.RS 4
Append files from given template directory to transfer tarball\&.
.RE
.PP
\fIfirewall_object_name\fR
.RS 4
Firewall object name\&.
.RE
.SH "EXAMPLES"
$ transfer_secuwall \-f /tmp/cluster\&.fwb \-d /tmp \-v /dev/sdc1 fw3
.sp
This will export the configuration of secunet wall firewall \fBfw3\fR to the partition \fB/dev/sdc1\fR, using the Firewall Builder XML file \fBcluster\&.fwb\fR in the working directory \fB/tmp\fR\&.
.sp
.SH "AUTHOR"
Written by Reto Buerki <buerki@swiss\-it\&.ch>\&.
.sp


doc/ChangeLog
2010-02-19  vadim  <vadim@vk.crocodile.org>

	* ProjectPanel_state_ops.cpp (ProjectPanel::loadState): refs #1236
	Crash on Mac. The GUI could be crashed by repetetivie scrolling of
	the tree and other parts of the main window with mouse scroll
	wheel on Mac

	* FindObjectWidget.cpp (FindObjectWidget::_findAll): fixed #1256
	search in the policy of firewalls does not work. This problem was
	introduced whith a fix for #1250

2010-02-15  yalovoy <yalovoy@gmail.com>
	* ProjectPanel.cpp: fixes #1244 GUI crash on exit
	Looks like it is sufficient to just open some data file, open rule set
	(I tried with Policy) and insert a rule in the middle. Then exist the
	program and it crashes.

2010-02-18  glitch.vk.crocodile.org  <vadim@vk.crocodile.org>

	* PolicyCompiler_iosacl.cpp (splitTCPServiceWithFlags::processNext):
	fixed #1247 incorrect use of tcp service with flags in
	object-group.  "object-group service" does not seem to support tcp
	flags and "established"

	* ObjectTreeViewItem.cpp (ObjectTreeViewItem::data):
	fixed #1248 Filtering in the object-tree crashes the GUI,
	SourceForge bug #2954501

	* ObjectManipulator_tree_ops.cpp (ObjectManipulator::updateObjectInTree):
	fixed #1249 GUI freezes when filters are applied. The GUI would
	freeze for a few seconds every time user modified anything in rules
	if a filter was applied to the tree at the same time.
	SourceForge bug #2954501

	* ObjectManipulator.cpp (ObjectManipulator::contextMenuRequested):
	fixed #1255 context menu items "Lock" and "Unlock" are not updated
	after object is locked.

	* ProjectPanel_events.cpp (ProjectPanel::event): fixed #1253
	Locking an object in the tree causes GUI crash.

2010-02-18  vadim  <vadim@vk.crocodile.org>

	* FWWindow.cpp (FWWindow::FWWindow): fixed #1252 The "New
	Object"-Dialog is always opened up twice when pressing <CTRL>+<N>

	* FindObjectWidget.cpp (FindObjectWidget::_findAll): fixed #1250
	The Find-Dialog searches 'Deleted Objects'-library although it's
	not enabled in the preferences. Sourceforge bug #2954501

	* IPv4Dialog.cpp (IPv4Dialog::DNSlookup): fixed #1251 DNS-Lookups
	in Address-Objects don't work. Sourceforge bug #2954501

2010-02-17  vadim  <vadim@vk.crocodile.org>

	* PolicyCompiler_iosacl.cpp (mirrorRule::processNext): fixes #1241
	"add mirror rule" checkbox for IOSACL. New feature: policy rule
	option "Add mirrored rule" (checkbox in the rule options dialog)
	makes policy compiler for IOS ACL automatically create a rule with
	mirrored source and destination addresses and service fields. This
	can be used to match "reply" packets using address and service
	parameters matched by this rule.  The action of the mirrored rule
	is the same as that of this one. Firewall Builder recognizes the
	following services and creates "mirrored" versions as follows:
	    - UDP service: mirrored service has source and destination port
	    ranges reversed
	    - TCP service: mirrored service has source and destination
	    port ranges reversed and "established" flag inverted.
	    If TCP service used in this rule does not have "established"
	    flag, the mirrored service gets it, and the other way around.
	    This is designed to simplify creating ACL rules to permit
	    "reply" TCP packets
	    - ICMP service: ICMP echo request is recognized, mirrored service
	    becomes ICMP echo reply. Other ICMP types are simply copied to the
	    mirrored service
	    - ICMPv6 service: like with ICMP, ICMP echo request is recognized
	    and other ICMPv6 types are just copied
	    - IP service: mirrored service is a copy

	* ssh_wrappers.cpp (ssh_wrapper): fixed #1246 policy installer
	fails on Linux. Policy installer failed with error "bash: -c: line
	6: syntax error: unexpected end of file" while trying to activate
	iptables script on Linux. This error was introduced in one of the
	recent builds.

2010-02-16  vadim  <vadim@vk.crocodile.org>

	* fixed #1243 "new cluster" operation creates failover groups that
	recursively refer to the interfaces of the same cluster. This
	happened when new cluster was created with a copy of rules from
	one of the firewalls.
	newClusterDialog_create.cpp

	* fixed #1235 "cluster with dynamic interface can not be used in
	rules". Fixes this and several other problems with policy and nat
	rules that use cluster inetrfaces with dynamic addresses.
	PolicyCompiler_ipt.cpp
	NATCompiler_ipt.cpp

	* fixed #1240 nat rule should get "on interface" when cluster
	interface is used in TSrc.
	NATCompiler_pf.cpp

	* Configlet.cpp (Configlet::expand): added basic protection
	against infinite loops in configlet expansion.

2010-02-15  vadim  <vadim@vk.crocodile.org>

	* Added template for the OpenWRT firewall. Fixes #1237

2010-02-15  glitch.vk.crocodile.org  <vadim@vk.crocodile.org>

	* working on memory leaks with the help of valgrind. Refs #1229
	Help.cpp
	Help.h
	RCS.cpp
	StartTipDialog.cpp
	StartTipDialog.h
	ssh_wrappers.cpp

2010-02-15  yalovoy <yalovoy@gmail.com>

	* refs #1202: Main menu "Rules" does not match rule context menu
	Items in the main menu Rules should get disabled and enabled just
	like items in the context menu do.
	Affected files:
	FWWindow.cpp
	RuleSetView.cpp
	RuleSetView.h


2010-02-14  glitch.vk.crocodile.org  <vadim@vk.crocodile.org>

	* ProjectPanel.cpp (ProjectPanel): refs #1229 memory leak problems.

2010-02-14  yalovoy <yalovoy@gmail.com>
	* fixes #1228: removing consecutive rules from a group in a ruleset
	its currently only possible to remove either the single top rule
	or single last rule from a group
	it would be useful to be able to select say the first several rules
	or last several rules and remove them all at once or if you select
	all rules in the group it would then just remove the group.
	Affected files:
	FWWindow.cpp
	FWWindow.h: #1228
	RuleSetView.cpp: #1228
	RuleSetView.h: #1228

2010-02-14  vadim  <vadim@vk.crocodile.org>

	* PolicyCompiler_ipt.cpp (PolicyCompiler_ipt::_expand_interface):

src/transfer_agents/secuwall/transfer_secuwall.cpp
/*
 * transfer_secuwall.cpp - secunet wall config transfer agent
 *
 * Copyright (c) 2008 secunet Security Networks AG
 * Copyright (c) 2008 Adrian-Ken Rueegsegger <rueegsegger@swiss-it.ch>
 * Copyright (c) 2008 Reto Buerki <buerki@swiss-it.ch>
 *
 * This work is dual-licensed under:
 *
 * o The terms of the GNU General Public License as published by the Free
 *   Software Foundation, either version 2 of the License, or (at your option)
 *   any later version.
 *
 * o The terms of NetCitadel End User License Agreement
 */

#include <iostream>
#include <fstream>
#include <sys/stat.h>

#ifdef _WIN32
#include <windows.h>
#include <direct.h>
#include <stdio.h>
#include <stdlib.h>
#endif

#ifdef HAVE_GETOPT_H
#include <getopt.h>
#else
#ifdef _WIN32
#include <getopt.h>
#else
#include <stdlib.h>
#endif
#endif

#include "../../../config.h"
#include "../../../build_num"

#include "fwbuilder/FWException.h"

#include "fwtransfer/TransferDevice.h"

// tarball base name and suffix (e.g. config.tar.gz)
#define BASENAME "config"
#define SUFFIX   "tar"

using namespace std;
using namespace fwtransfer;
using namespace libfwbuilder;

int fwbdebug = 0;

// object database filename: not used at the moment
static string filename     = "";
// workdir of firewall to export
static string workdir      = "";
// templates dir to use
static string tmpldir      = "";
// volumeid (e.g. /dev/sdc1) to use as transfer partition
static string volumeid     = "";
// object name of firewall to export (e.g. fw1)
static string fwobjectname = "";

// append fwobjectname to transfer tarball
static bool appendname     = false;

/**
 * init portable usb device list.
 *
 * @param devices device list to init, only portable devices are valid.
 * @return true if successful, false if not
 */
bool init_usbdisks(TransferDeviceList &devices)
{
    try
    {
        devices.init();
    }
    catch (FWException &ex)
    {
        cout << "Could not init list of usbdisks!" << endl;
        cout << "Error: " << ex.toString() << endl;
        return false;
    }

    TransferDeviceList::const_iterator it;
    it = devices.begin();
    if (it == devices.end())
    {
        cout << "No usable transfer volumes found! " << endl;
        return false;
    }

    return true;
}

/**
 * display transfer_secuwall usage.
 */
void usage(const char *name)
{
    cout << "Firewall Builder: config transfer agent for "
        "secunet wall host OS" << endl;
    cout << ("Version : ") << VERSION << endl;
    cout << ("Usage   : ") << name <<
        " [-l] [-h] [-n] -v volumeid [-f filename.xml] [-d workdir] "
        "[-a templatedir] firewall_object_name" << endl;
}

/**
 * list all available transfer volumes.
 */
void list_volumes()
{
    TransferDeviceList devices;

    if (!init_usbdisks(devices))
    {
        exit(EXIT_FAILURE);
    }

    TransferDeviceList::const_iterator it;
    it = devices.begin();
    cout << endl;
    for (; it != devices.end(); it++)
    {
        TransferDevice dev = *it;
        dev.dump();
    }

    exit(EXIT_SUCCESS);
}

/**
 * copy file specified by src to dst
 *
 * @param src source filepath
 * @param dst destination filepath
 * @return true if successfully copied, false if not
 */
bool copy_file(const string &src, const string &dst)
{
    std::ifstream infile(src.c_str(), std::ios_base::binary);
    std::ofstream outfile(dst.c_str(), std::ios_base::binary);

    if (!infile)
    {
        cout << "File not found: " << src << endl;
        return false;
    }

    if (!outfile)
    {
        cout << "Could not create: " << dst << endl;
        return false;
    }

    try
    {
        outfile << infile.rdbuf();
    }
    catch (...)
    {
        cout << "Error while writing to: " << dst << endl;
        infile.close();
        outfile.close();
        return false;
    }
    infile.close();
    outfile.close();
    return true;
}

/**
 * secunet wall config transfer agent
 *
 * used to export fw config tarball to a portable usb medium.
 */
int main(int argc, char **argv)
{
    if (argc <= 1)
    {
        usage(argv[0]);
        exit(EXIT_FAILURE);
    }

    int opt;

    while ((opt = getopt(argc, argv, "lhnv:f:d:a:")) != EOF)
    {
        switch (opt)
        {
        case 'd':
            workdir = string(optarg);
            break;
        case 'a':
            tmpldir = string(optarg);
            break;
        case 'f':
            filename = string(optarg);
            break;
        case 'v':
            volumeid = string(optarg);
            break;
        case 'l':
            list_volumes();
            break;
        case 'h':
            usage(argv[0]);
            exit(EXIT_FAILURE);
        case 'n':
            appendname = true;
            break;
        default:
            usage(argv[0]);
            exit(EXIT_FAILURE);
        }
    }

    if ((argc - 1) != optind)
    {
        usage(argv[0]);
        exit(EXIT_FAILURE);
    }

    fwobjectname = string(argv[optind++]);

    if (workdir.empty())
    {
        workdir = "./";
    }

    if (filename.empty())
    {
        filename = workdir + fwobjectname + ".fwb";
    }

    if (fwbdebug)
    {
        cout << "Volume\t: "  << volumeid << endl;
        cout << "Object\t: "  << fwobjectname << endl;
        cout << "Workdir\t: " << workdir << endl;
        cout << "DB\t: "      << filename << endl;
    }

    if (
#ifdef _WIN32
        _chdir(workdir.c_str())
#else
        chdir(workdir.c_str())
#endif
    )
    {
        cout << "Can't change directory to: " << workdir << endl;
        exit(EXIT_FAILURE);
    }

    // check for existence of fwobjectname subdir
    struct stat buffer;
    if (stat(fwobjectname.c_str(), &buffer))
    {
        cout << "Config subdir not found for " << fwobjectname << endl;
        exit(EXIT_FAILURE);
    }

    // check templates directory (with -a flag only)
    if (!tmpldir.empty())
    {
        if (stat(tmpldir.c_str(), &buffer))
        {
            cout << "Templates directory '" << tmpldir << "' not found" << endl;
            exit(EXIT_FAILURE);
        }
    }

    // try to mount volume
    TransferDeviceList devices;
    if (!init_usbdisks(devices))
    {
        exit(EXIT_FAILURE);
    }

    TransferDeviceList::const_iterator it;
    it = devices.getDeviceByName(volumeid);
    if (it == devices.end())
    {
        cout << "Could not find volume " << volumeid << endl;
        exit(EXIT_FAILURE);
    }
    TransferDevice dev = *it;

    if (!dev.isMounted())
    {
        try
        {
            dev.mount();
        }
        catch (FWException &ex)
        {
            cout << "Could not mount volume " << volumeid << endl;
            cout << "Error: " << ex.toString() << endl;
            exit(EXIT_FAILURE);
        }
    }
    string mountpoint = dev.getMountpoint().toStdString();
    cout << "Device mounted to " << mountpoint << endl;

    // copy firewall script to fwobjectname subdir, we assume the script name
    // to be: 'fwobjectname + fw suffix'. the file will be copied to 'sysconfig'
    // subdirectory.
    string fwscriptname = fwobjectname + ".fw";
    string fwscriptdst = fwobjectname + "/sysconfig/" + fwscriptname;
    if (!copy_file(fwscriptname, fwscriptdst))
    {
        cout << "Firewall config script file not found for '"
             << fwobjectname << "'" << endl;
        exit(EXIT_FAILURE);
    }

    // construct tarball name depending on appendname flag
    string tarball;
    if (appendname)
    {
        tarball = string(BASENAME) + "-" + fwobjectname + "." + string(SUFFIX);
    }
    else
    {
        tarball = string(BASENAME) + "." + string(SUFFIX);
    }

    // tell tar to change into fwobjectname subdir and create tarball
    // force file ownership to root:root
    string cmd = "tar cCf " + fwobjectname + " " + tarball + " --owner=0 --group=0 .";
    if (system(cmd.c_str()) != 0)
    {
        cout << "Could not create tarball '" << tarball << "'" << endl;
        exit(EXIT_FAILURE);
    }

    // append templates to tarball
    if (!tmpldir.empty())
    {
        cout << "Adding templates from '" << tmpldir << "'" << endl;
        cmd = "tar rCf " + tmpldir + " " + tarball + " --owner=0 --group=0 .";
        if (system(cmd.c_str()) != 0)
        {
            cout << "Could not append templates to tarball '" << tarball
                 << "'" << endl;
            exit(EXIT_FAILURE);
        }
    }

    // compress tarball
    cmd = "gzip -f " + tarball;
    if (system(cmd.c_str()) != 0)
    {
        cout << "Could not compress tarball '" << tarball << "'" << endl;
        cout << "Missing gzip binary?" << endl;
        exit(EXIT_FAILURE);
    }

    // copy compressed archive to volume
    string compressed = tarball + ".gz";
    string outpath = mountpoint + "/" + compressed;
    if (!copy_file(compressed, outpath))
    {
        cout << "Could not copy '" << compressed << "' to '"
             << outpath << "'" << endl;
        exit(EXIT_FAILURE);
    }

#ifndef _WIN32
    sync();
#endif

    cout << "Exported '" << fwobjectname << "' config to " << outpath << endl;
    exit(EXIT_SUCCESS);
}
